src/pfe/portal/modules/Templates.js

```diff
@@ -109,67 +127,152 @@ module.exports = class Templates {
       log.info(`Repository list updated.`);
-    });
+    } catch (err) {
+      log.error(`Error writing repository list to ${this.repositoryFile}: ${err}`)
+    }
   }
 
   getRepositories() {
     return this.repositoryList;
   }
 
+  getEnabledRepositories() {
+    return this.getRepositories().filter(repo =>
+      // if the repo doesn't specify whether it's enabled, consider it enabled
+      (repo.enabled || !repo.hasOwnProperty('enabled'))
+    );
+  }
+
+  doesRepositoryExist(repoUrl) {
+    try {
+      this.getRepository(repoUrl);
+      return true;
+    } catch (error) {
+      return false;
+    }
+  }
+
+  getRepositoryIndex(url) {
+    const repos = this.getRepositories();
+    const index = repos.findIndex(repo => repo.url === url);
+    return index;
+  }
+
+  /**
+   * @param {String} url
+   * @return {JSON} reference to the repo object in this.repositoryList
+   */
+  getRepository(url) {
+    const index = this.getRepositoryIndex(url);
+    if (index < 0) throw new Error(`no repository found with URL '${url}'`);
+    const repo = this.getRepositories()[index];
+    return repo;
+  }
+
+  async enableRepository(url) {
+    const repo = this.getRepository(url);
+    repo.enabled = true;
+    await this.writeRepositoryList();
+
+  }
+
+  async disableRepository(url) {
+    const repo = this.getRepository(url);
+    repo.enabled = false;
+    await this.writeRepositoryList();
+  }
+
+  async performOperation(operation) {
+    const { op, path, value } = operation;
+    let operationResult = {};
+    if (op === 'enable') {
+      operationResult = await this.performEnableOrDisableOperation({ path, value });
+    }
+    operationResult.requestedOperation = operation;
+    return operationResult;
+  }
+
+  /**
+   *
+   * @returns {JSON} { status, error (optional) }
+   */
+  async performEnableOrDisableOperation({ path, value }) {
+    const repoUrl = path;
+    if (!this.doesRepositoryExist(repoUrl)) {
+      return {
+        status: 404,
+        error: 'Unknown repository URL',
+      };
+    }
+    try {
+      if (value === 'true') {
+        await this.enableRepository(repoUrl);
+      } else {
+        await this.disableRepository(repoUrl);
+      }
+      return {
+        status: 200
+      };
+    } catch (error) {
+      return {
+        status: 500,
+        error: error.message,
+      };
+    }
+  }
+
   /**
    * Add a repository to the list of template repositories.
    *
    * @param {*} url - url of the template repository
    * @param {*} description - description of the template repository
    */
-  addRepository(repositoryUrl, repositoryDescription) {
-    if (this.getRepositories().filter(repo => repo.url == repositoryUrl).length != 0) {
-      return false;
+  async addRepository(repositoryUrl, repositoryDescription) {
+    if (this.getRepositories().find(repo => repo.url == repositoryUrl)) {
+      throw new Error('Repository URL must be unique');
```

Comment:
Similar to above comment, why is this changed to throwing error?

In this case, this would actually affect `ExtensionList` here https://github.com/eclipse/codewind/pull/122/files#diff-2e95eceb73a0346dfd44f09698ce42b3R45 because if any extension supplies a duplicate repo url and the add throws an error, it stops other extensions from loading (but there's a flaw in the initialization to begin with, please comment above)

Reply:
(As above) I changed `addRepository` to throw an error because I thought that being unable to add a repo is exceptional behaviour. I thought it best practice to throw an error rather than returning false, so that we get explicit information about why the call failed.

I'm happy to use (as you suggested) a try-catch in ExtensionList so that individual failures do not stop other extensions from loading